UIforETW/UIforETWDlg.cpp

```diff
@@ -40,10 +41,821 @@ const int kRecordTraceHotKey = 1234;
+	const int addStringResult = comboBox->AddString(stringToAdd);
+	if (addStringResult == CB_ERR)
+	{
+		MessageBoxW(NULL, stringToAdd, L"Unexpected (fatal) error adding string to comboBox!", MB_OK);
+		debug::Alias(&stringToAdd);
+		std::terminate();
+	}
+	if (addStringResult == CB_ERRSPACE)
+	{
+		MessageBoxW(NULL, stringToAdd, L"Not enough space available to store string!", MB_OK);
+		std::terminate();
+	}
+	OutputDebugStringA("UIforETW: successfully added string `");
+	OutputDebugStringW(stringToAdd);
+	OutputDebugStringA("` to CComboBox...\r\n");
+}
+
+void addbtInputTracingStrings( _Inout_ CComboBox* const btInputTracing )
+{
+	addStringToCComboBox(btInputTracing, L"Off");
+	addStringToCComboBox(btInputTracing, L"Private");
+	addStringToCComboBox(btInputTracing, L"Full");
+}
+
+void addbtTracingModeStrings( _Inout_ CComboBox* const btTracingMode )
+{
+	addStringToCComboBox(btTracingMode, L"Circular buffer tracing");
+	addStringToCComboBox(btTracingMode, L"Tracing to file");
+	addStringToCComboBox(btTracingMode, L"Heap tracing to file");
+}
+
+_Success_( return )
+bool setCComboBoxSelection( _Inout_ CComboBox* const comboBoxToSet, _In_ _In_range_( -1, INT_MAX ) const int selectionToSet )
+{
+	//CComboBox::SetCurSel calls SendMessage, to send a CB_SETCURSEL message.
+
+	//The documentation for CB_SETCURSEL is a bit confusing!
+	//CB_SETCURSEL, if successful, returns the index of the item that was selected.
+	//CB_SETCURSEL, on failure, returns CB_ERR.
+	//The documentation suggests that -1 might be a valid argument for clearing the selection,
+	//however, it also says that if passed -1, it will return CB_ERR.
+	const int setSelectionResult = comboBoxToSet->SetCurSel(selectionToSet);
+	if (setSelectionResult == CB_ERR)
+	{
+		outputPrintf(L"Failed to set selection (for a comboBox)\n"
+					 L"\tIndex that we attempted to set the selection to: %i\n"
+					 L"\tCue banner of CComboBox: %s\n",
+					 selectionToSet, comboBoxToSet->GetCueBanner().GetString());//Grr. I hate CString.
+		return false;
+	}
+	return true;
+}
+
+template<class DerivedButton>
+void addSingleToolToToolTip(_Inout_ CToolTipCtrl*  const toolTip,
+							_Inout_ DerivedButton* const btToAdd,
+							_In_z_  PCWSTR         const toolTipMessageToAdd )
+{
+	static_assert( std::is_base_of<CWnd, DerivedButton>::value,
+				   "Cannot add a non-CWnd-derived class as a tooltip!" );
+
+	const BOOL addToolTipResult = toolTip->AddTool(btToAdd, toolTipMessageToAdd);
+	
+	if (addToolTipResult == TRUE)
+		return;
+	OutputDebugStringA("UIforETW: Failed to add a message to a tooltip!!\r\n");
+	OutputDebugStringA("The tooltip message:\r\n\t`");
+	OutputDebugStringW(toolTipMessageToAdd);
+	OutputDebugStringA("`\r\n");
+	std::terminate();
+}
+
+
+CRect GetWindowRectFromHwnd( _In_ const HWND hwnd )
+{
+	RECT windowRect_temp;
+	ASSERT(::IsWindow(hwnd));
+	const BOOL getWindowRectResult = ::GetWindowRect(hwnd, &windowRect_temp);
+	if (getWindowRectResult == 0)

```

Comment:
Fair enough, although I don't think this is an *impossible* error.

TLDR: Today I'm running with -Wpedantic, -Wnitpick, and -Wannoying. Will remove this code.

* It can happen, and I found [one (if only one) instance of problems caused by not checking the return value](http://forums.codeguru.com/printthread.php?t=305640)
* There is a minute chance of a race-condition, where the window is destroyed (and thus the handle is invalid) before `GetWindowRect` is called - in which case maybe we should mark the function with something like `_No_competing_thread_`? (strangely, that annotation isn't documented)
* Unlike `wcscpy_s`, we don't have the source code, so by assuming that failure is impossible is technically an implementation detail, and as a longtime reader of Raymond Chen's blog, I have ethical issues with relying on implementation details.